switch to difflib.diff_bytes in python3 case.

lethliel opened the pull request:
The files are now opened as rb for diffing. In python2 nothing changes.
In python3 the returned diff is bytes now.

The following changes were made:
  * commandline.py: The returned diff is now bytes
  * get_diff now returs the diff as a bytes-like object
  * run_pager writes with sys.stdout.buffer.write if message
    is not a string
  * for the commit message the returned diff needs to be decoded
    now. Otherwise it will just producce garbage. For the commit
    message the diff on decoded bytes-objects is ok. (nothing harmfull
    can happen here)

This is the alternative to https://github.com/openSUSE/osc/pull/554 and does not decode the generated 
bytes-diff.

lethliel commented:
@marcus-h This is an alternating approach to https://github.com/openSUSE/osc/issues/540 which does not decode the diff. 

The falling testsuite for 3.3 and 3.4 is the following error: 

`unsupported operand type(s) for %: 'bytes' and 'bytes'` but I will disable the testsuite for 3.3 and 3.4 anyway. 

Can you please review.

marcus-h commented:
Hi,

just a very very quick and high-level review:

On 2019-07-17 01:45:51 -0700, Marco Strigl wrote:
> From b3bcfac72c521d55e668319c441d6ccb5775342d Mon Sep 17 00:00:00 2001

> Date: Wed, 17 Jul 2019 10:34:20 +0200
> Subject: [PATCH 1/4] switch to difflib.diff_bytes in python3 case.
>
> The files are now opened as rb for diffing. In python2 nothing changes.
> In python3 the returned diff is bytes now.
>
> The following changes were made:
>   * commandline.py: The returned diff is now bytes
>   * get_diff now returs the diff as a bytes-like object
>   * run_pager writes with sys.stdout.buffer.write if message
>     is not a string
>   * for the commit message the returned diff needs to be decoded
>     now. Otherwise it will just producce garbage. For the commit
>     message the diff on decoded bytes-objects is ok. (nothing harmfull
>     can happen here)
> ---
>  osc/commandline.py | 16 ++++++----
>  osc/core.py        | 78 +++++++++++++++++++++++++---------------------
>  2 files changed, 53 insertions(+), 41 deletions(-)
>
> diff --git a/osc/commandline.py b/osc/commandline.py
> index 8ea37ba3..85ee1a34 100644
> --- a/osc/commandline.py
> +++ b/osc/commandline.py
> @@ -3864,15 +3864,16 @@ def do_diff(self, subcmd, opts, *args):
>                  return
>          else:
>              rev1, rev2 = parseRevisionOption(opts.revision)
> -        diff = ''

Hmm why? This breaks the old behavior.

>          for pac in pacs:
> +            diff = b''

See above.

>              if not rev2:
>                  for i in pac.get_diff(rev1):
> -                    diff += ''.join(i)
> +                    diff += b''.join(i)
>              else:
> -                diff += decode_it(server_diff_noex(pac.apiurl, pac.prjname, pac.name, rev1,
> -                                    pac.prjname, pac.name, rev2,
> -                                    not opts.plain, opts.missingok, opts.meta, not opts.unexpand))
> +                diff = b''

See above.

> +                diff += server_diff_noex(pac.apiurl, pac.prjname, pac.name, rev1,
> +                                pac.prjname, pac.name, rev2,
> +                                not opts.plain, opts.missingok, opts.meta, not opts.unexpand)
>          run_pager(diff)
>
>
> @@ -3976,7 +3977,10 @@ def do_rdiff(self, subcmd, opts, *args):
>                              expand=not opts.unexpand,
>                              onlyissues=opts.issues_only)
>          if opts.issues_only:
> -            print(rdiff)
> +            if isinstance(rdiff, str):
> +                print(rdiff)
> +            else:
> +                sys.stdout.buffer.write(rdiff)
>          else:
>              run_pager(rdiff)
>
> diff --git a/osc/core.py b/osc/core.py
> index 23ca55f6..be77b2ee 100644
> --- a/osc/core.py
> +++ b/osc/core.py
> @@ -1963,26 +1963,26 @@ def status(self, n):
>
>      def get_diff(self, revision=None, ignoreUnversioned=False):
>          import tempfile
> -        diff_hdr = 'Index: %s\n'
> -        diff_hdr += '===================================================================\n'
> +        diff_hdr = b'Index: %s\n'
> +        diff_hdr +=b'===================================================================\n'
>          kept = []
>          added = []
>          deleted = []
>          def diff_add_delete(fname, add, revision):
>              diff = []
> -            diff.append(diff_hdr % fname)
> +            diff.append(diff_hdr % fname.encode())
>              tmpfile = None
>              origname = fname
>              if add:
> -                diff.append('--- %s\t(revision 0)\n' % fname)
> +                diff.append(b'--- %s\t(revision 0)\n' % fname.encode())
>                  rev = 'revision 0'
>                  if revision and not fname in self.to_be_added:
>                      rev = 'working copy'
> -                diff.append('+++ %s\t(%s)\n' % (fname, rev))
> +                diff.append(b'+++ %s\t(%s)\n' % (fname.encode(), rev.encode()))
>                  fname = os.path.join(self.absdir, fname)
>              else:
> -                diff.append('--- %s\t(revision %s)\n' % (fname, revision or self.rev))
> -                diff.append('+++ %s\t(working copy)\n' % fname)
> +                diff.append(b'--- %s\t(revision %s)\n' % (fname.encode(), revision.encode() or self.rev.encode()))

revision might be None.

> +                diff.append(b'+++ %s\t(working copy)\n' % fname.encode())
>                  fname = os.path.join(self.storedir, fname)
>
>              try:
> @@ -1991,22 +1991,22 @@ def diff_add_delete(fname, add, revision):
>                      get_source_file(self.apiurl, self.prjname, self.name, origname, tmpfile, revision)
>                      fname = tmpfile
>                  if binary_file(fname):
> -                    what = 'added'
> +                    what = b'added'
>                      if not add:
> -                        what = 'deleted'
> +                        what = b'deleted'
>                      diff = diff[:1]
> -                    diff.append('Binary file \'%s\' %s.\n' % (origname, what))
> +                    diff.append(b'Binary file \'%s\' %s.\n' % (origname.encode(), what))
>                      return diff
> -                tmpl = '+%s'
> -                ltmpl = '@@ -0,0 +1,%d @@\n'
> +                tmpl = b'+%s'
> +                ltmpl = b'@@ -0,0 +1,%d @@\n'
>                  if not add:
> -                    tmpl = '-%s'
> -                    ltmpl = '@@ -1,%d +0,0 @@\n'
> -                lines = [tmpl % i for i in open(fname, 'r').readlines()]
> +                    tmpl = b'-%s'
> +                    ltmpl = b'@@ -1,%d +0,0 @@\n'
> +                lines = [tmpl % i for i in open(fname, 'rb').readlines()]
>                  if len(lines):
>                      diff.append(ltmpl % len(lines))
> -                    if not lines[-1].endswith('\n'):
> -                        lines.append('\n\\ No newline at end of file\n')
> +                    if not lines[-1].endswith(b'\n'):
> +                        lines.append(b'\n\\ No newline at end of file\n')

Hmm why not simply keep the str constructions here and encode
it later?

>                  diff.extend(lines)
>              finally:
>                  if tmpfile is not None:

<SNIP>

> @@ -7253,7 +7261,7 @@ def get_commit_message_template(pac):
>              f.close()
>
>      if diff:
> -        template = parse_diff_for_commit_message(''.join(diff))
> +        template = parse_diff_for_commit_message(''.join(decode_list(diff)))
>
Ok.

>      return template
>
> @@ -7288,7 +7296,7 @@ def get_commit_msg(wc_dir, pacs):
>          if changed:
>              footer += changed
>              footer.append('\nDiff for working copy: %s' % p.dir)
> -            footer.extend([''.join(i) for i in p.get_diff(ignoreUnversioned=True)])
> +            footer.extend([''.join(decode_list(i)) for i in p.get_diff(ignoreUnversioned=True)])

Ok.

>              lines.extend(get_commit_message_template(p))
>      if template is None:
>          if lines and lines[0] == '':
>
> From 5686c7f52283463acc394d610ac0caf0508b6a62 Mon Sep 17 00:00:00 2001

> Date: Wed, 17 Jul 2019 11:08:20 +0200
> Subject: [PATCH 2/4] do not try to decode None Type and Int
>
> ---
>  osc/core.py | 6 +++++-
>  1 file changed, 5 insertions(+), 1 deletion(-)
>
> diff --git a/osc/core.py b/osc/core.py
> index be77b2ee..0f37916d 100644
> --- a/osc/core.py
> +++ b/osc/core.py
> @@ -1981,7 +1981,11 @@ def diff_add_delete(fname, add, revision):
>                  diff.append(b'+++ %s\t(%s)\n' % (fname.encode(), rev.encode()))
>                  fname = os.path.join(self.absdir, fname)
>              else:
> -                diff.append(b'--- %s\t(revision %s)\n' % (fname.encode(), revision.encode() or self.rev.encode()))
> +                if revision:
> +                    revision = str(revision).encode()
> +                else:
> +                    revision = str(self.rev).encode()

self.rev is (always) a str.
revision is a bytes and used afterwards, does this work?

Hmm can you squash all commits? (I think I commented on this in a previous
patch.)

> +                diff.append(b'--- %s\t(revision %s)\n' % (fname.encode(), revision))
>                  diff.append(b'+++ %s\t(working copy)\n' % fname.encode())
>                  fname = os.path.join(self.storedir, fname)
>
>

marcus-h commented:
We should also fix up:
- submit_action_diff (it should always return bytes)
- request_interactive_review (which calls submit_action_diff)

(In both cases, the fixes should be straightforward.)

lethliel commented:
@marcus-h: I squashed all commits :-)

lethliel commented:
@marcus-h: 

I reworked the PR based on your comments. 
Just one remark: 

I do not keep the str constructions because they are all put in a list. I find it easier to use bytes here directly than to encode the complete list at the end.

marcus-h commented:
On 2019-07-17 13:47:03 +0000, Marco Strigl wrote:
> @marcus-h:
>
> I reworked the PR based on your comments.
> Just one remark:
>
> I do not keep the str constructions because they are all put in a list. I find it easier to use bytes here directly than to encode the complete list at the end.
>
Ok. Fine with me:)

> diff --git a/osc/commandline.py b/osc/commandline.py
> index 8ea37ba3..ff32b42e 100644
> --- a/osc/commandline.py
> +++ b/osc/commandline.py
> @@ -3976,7 +3976,10 @@ def do_rdiff(self, subcmd, opts, *args):
>                              expand=not opts.unexpand,
>                              onlyissues=opts.issues_only)
>          if opts.issues_only:
> -            print(rdiff)
> +            if isinstance(rdiff, str):
> +                print(rdiff)
> +            else:
> +                sys.stdout.buffer.write(rdiff)

Hmm this is not needed because in the issues_only case a str is
returned. Either we omit this change or the "onlyissues" codepath
in server_diff should return a bytes.

>          else:
>              run_pager(rdiff)
>
> diff --git a/osc/core.py b/osc/core.py
> index 23ca55f6..5f662276 100644
> --- a/osc/core.py
> +++ b/osc/core.py
> @@ -1963,26 +1963,30 @@ def status(self, n):
>
>      def get_diff(self, revision=None, ignoreUnversioned=False):
>          import tempfile
> -        diff_hdr = 'Index: %s\n'
> -        diff_hdr += '===================================================================\n'
> +        diff_hdr = b'Index: %s\n'
> +        diff_hdr +=b'===================================================================\n'

Missing whitespace.

<SNIP>

> @@ -4795,22 +4799,30 @@ def get_source_file_diff(dir, filename, rev, oldfilename = None, olddir = None,
>          if f2:
>              f2.close()
>
> -    d = difflib.unified_diff(s1, s2,
> -        fromfile = '%s\t(revision %s)' % (origfilename, rev), \
> +    from_file = b'' + origfilename.encode() + b'\t(revision ' + str(rev).encode() + b')'
> +    to_file = b'' + origfilename.encode() + b'\t(working copy)'

Why the leading b''? Can't we simply use %-interpolation to construct
the bytes? That is,

from_file = b'%s\t(revision %s)' % (origfilename.encode(), str(rev.encode))
to_file = ...

> +
> +    if sys.version_info < (3,0):
> +        d = difflib.unified_diff(s1, s2,
> +            fromfile = '%s\t(revision %s)' % (origfilename, rev), \
>          tofile = '%s\t(working copy)' % origfilename)

Hmm just pass from_file and to_file from above?

> +    else:
> +        d = difflib.diff_bytes(difflib.unified_diff, s1, s2, \
> +            fromfile = from_file, \
> +            tofile = to_file)

<SNIP>

> @@ -7438,12 +7450,12 @@ def get_formatted_issues(apiurl, reqid):
>                      tmpfile.close()
>                      tmpfile = None
>                  if tmpfile is None:
> -                    tmpfile = tempfile.NamedTemporaryFile(suffix='.diff', mode='r+')
> -                    tmpfile.write(req_summary)
> -                    tmpfile.write(issues)
> +                    tmpfile = tempfile.NamedTemporaryFile(suffix='.diff', mode='rb+')
> +                    tmpfile.write(req_summary.encode())
> +                    tmpfile.write(issues.encode())
>                      try:
>                          diff = request_diff(apiurl, request.reqid)
> -                        tmpfile.write(decode_it(diff))
> +                        tmpfile.write(diff)
>                      except HTTPError as e:
>                          if e.code != 400:
>                              raise

The "diff = 'old: %s/%s\nnew:..." part in the except block should be
a bytes, too (and the subsequent "diff += '\n\n'").

lethliel commented:
@marcus-h: 

* using %-interpolation to create from_file and to_file and reuse it in the python2 case.
* removed unused if in issues_only and added whitespace. 
* fixed encoding in except block

marcus-h commented:
On 2019-07-19 00:06:15 -0700, Marco Strigl wrote:
> @@ -4795,22 +4799,34 @@ def get_source_file_diff(dir, filename, rev, oldfilename = None, olddir = None,
>          if f2:
>              f2.close()
>
> -    d = difflib.unified_diff(s1, s2,
> -        fromfile = '%s\t(revision %s)' % (origfilename, rev), \
> -        tofile = '%s\t(working copy)' % origfilename)
> +    from_file = b'' + origfilename.encode() + b'\t(revision ' + str(rev).encode() + b')'
> +    to_file = b'' + origfilename.encode() + b'\t(working copy)'
> +

Let's remove the assignments from above:)

> +
> +    from_file = b'%s\t(revision %s)' % (origfilename.encode(), str(rev).encode())
> +    to_file = b'%s\t(working copy)' % origfilename.encode()
> +

<SNIP>

> @@ -7438,21 +7454,21 @@ def get_formatted_issues(apiurl, reqid):
>                      tmpfile.close()
>                      tmpfile = None
>                  if tmpfile is None:
> -                    tmpfile = tempfile.NamedTemporaryFile(suffix='.diff', mode='r+')
> -                    tmpfile.write(req_summary)
> -                    tmpfile.write(issues)
> +                    tmpfile = tempfile.NamedTemporaryFile(suffix='.diff', mode='rb+')
> +                    tmpfile.write(req_summary.encode())
> +                    tmpfile.write(issues.encode())
>                      try:
>                          diff = request_diff(apiurl, request.reqid)
> -                        tmpfile.write(decode_it(diff))
> +                        tmpfile.write(diff)
>                      except HTTPError as e:
>                          if e.code != 400:
>                              raise
>                          # backward compatible diff for old apis
>                          for action in src_actions:
> -                            diff = 'old: %s/%s\nnew: %s/%s\n' % (action.src_project, action.src_package,
> +                            diff = b'old: %s/%s\nnew: %s/%s\n' % (action.src_project, action.src_package,
>                                  action.tgt_project, action.tgt_package)

Hmm we have to/should encode the *_project and *_package attributes
as well.

marcus-h commented:
Some final remarks:
- in get_commit_message_template we should open the file with
  mode 'rb' (in the elif block) and '+' should be a b'+'

- in do_submitrequest we have:

        rdiff = None
        if opts.diff or not opts.message:
            try:
                rdiff = 'old: %s/%s\nnew: %s/%s rev %s\n' % (dst_project, dst_package, src_project, src_package, rev)
                rdiff += decode_it(server_diff(apiurl,
                                    dst_project, dst_package, None,
                                    src_project, src_package, rev, True))
            except:
                rdiff = ''

        if opts.diff:
            run_pager(rdiff)
            return

  We should convert everything to bytes and do the decoding after the
  return (because then it is "just" used for the sr msg). That is,

  rdiff = None
  ...
  if opts.diff:
      run_pager(rdiff)
      return
  if rdiff is not None:
      rdiff = decode_it(rdiff)

- in _submit_request we have:

                    if t:
                        rdiff = ''
                        try:
                            rdiff = server_diff(apiurl, t, p, opts.revision, project, p, None, True)
                        except:
                            rdiff = ''

                        if rdiff != '':
                            targetprojects.append(t)
                            pac.append(p)
                        else:

  all '' should be replaced with b''

- in _prdiff_output_diff we should not decode the rdiff in the else
  block. We should probably do something similar as we already do in
  run_pager.

- submit_action_diff has to be converted to bytes, too


From the current PR:

> @@ -4868,12 +4880,12 @@ def server_diff_noex(apiurl,
>          try:
>              body = decode_it(e.read())
>              if not 'bad link' in body:
> -                return '# diff failed: ' + body
> +                return b'# diff failed: ' + body

This will fail because body is a str. Let's simply do

body = e.read()
if not b'bad link' in body:
    ...

>          except:
> -            return '# diff failed with unknown error'
> +            return b'# diff failed with unknown error'
>
>          if expand:
> -            rdiff =  "## diff on expanded link not possible, showing unexpanded version\n"
> +            rdiff =  b"## diff on expanded link not possible, showing unexpanded version\n"
>              try:
>                  rdiff += server_diff_noex(apiurl,
>                      old_project, old_package, old_revision,
> @@ -4884,7 +4896,7 @@ def server_diff_noex(apiurl,
>                  summary = ''
>                  if not elm is None:
>                      summary = elm.text
> -                return 'error: diffing failed: %s' % summary
> +                return b'error: diffing failed: %s' % summary

Hmm we should probably encode() summary.

>              return rdiff

marcus-h commented:
On 2019-07-22 06:50:17 -0700, Marco Strigl wrote:
> diff --git a/osc/commandline.py b/osc/commandline.py
> index 8ea37ba3..86ef1b46 100644
> --- a/osc/commandline.py
> +++ b/osc/commandline.py
> @@ -1405,16 +1405,19 @@ def _check_service(root):
>          rdiff = None
>          if opts.diff or not opts.message:
>              try:
> -                rdiff = 'old: %s/%s\nnew: %s/%s rev %s\n' % (dst_project, dst_package, src_project, src_package, rev)
> -                rdiff += decode_it(server_diff(apiurl,
> -                                    dst_project, dst_package, None,
> -                                    src_project, src_package, rev, True))
> +                rdiff = b'old: %s/%s\nnew: %s/%s rev %s\n' % (dst_project, dst_package, src_project, src_package, rev)

We should also encode() the arguments.

> @@ -4156,7 +4159,13 @@ def _prdiff_output_diff(self, opts, rdiff):
>              print("".join(decode_it(x) for x in p.stdout.readlines()))
>          elif opts.unified:
>              print()
> -            print(decode_it(rdiff))
> +            if isinstance(rdiff, str):
> +                print(rdiff)
> +            else:
> +                try:
> +                    sys.stdout.buffer.write(rdiff)
> +                except AttributeError as e:
> +                    print(decode_it(rdiff))

Hmm I'm not really fond of it but OK for now.

> diff --git a/osc/core.py b/osc/core.py
> index 23ca55f6..4b1fa793 100644
> --- a/osc/core.py
> +++ b/osc/core.py
> @@ -4884,7 +4896,7 @@ def server_diff_noex(apiurl,
>                  summary = ''
>                  if not elm is None:
>                      summary = elm.text
> -                return 'error: diffing failed: %s' % summary
> +                return b'error: diffing failed: %s' % summary.encode()
>              return rdiff
>
>
> @@ -4933,10 +4945,10 @@ def submit_action_diff(apiurl, action):
>                  if e.code != 404:
>                      raise e
>                  root = ET.fromstring(e.read())
> -                return 'error: \'%s\' does not exist' % root.find('summary').text
> +                return b'error: \'%s\' does not exist' % root.find('summary').text

This won't work because .text is a str (missing encode()).
[Actually, we should be careful here and check if a summary node exists etc.
but well... the old code didn't do that either:/ ]

>          elif e.code == 404:
>              root = ET.fromstring(e.read())
> -            return 'error: \'%s\' does not exist' % root.find('summary').text
> +            return b'error: \'%s\' does not exist' % root.find('summary').text

As above.

>  def make_dir(apiurl, project, package, pathname=None, prj_dir=None, package_tracking=True, pkg_path=None):
> @@ -7247,13 +7259,13 @@ def get_commit_message_template(pac):
>          if pac.status(filename) == 'M':
>              diff += get_source_file_diff(pac.absdir, filename, pac.rev)
>          elif pac.status(filename) == 'A':
> -            f = open(os.path.join(pac.absdir, filename), 'r')
> +            f = open(os.path.join(pac.absdir, filename), 'rb')
>              for line in f:
> -                diff += '+' + line
> +                diff += b'+' + line

Hmm I just realized that this won't do the correct thing in case
of bytes. We should probably do something like

diff.extend((b'+' + line for line in f))
or
diff.extend((b'+' + line for line in f.readlines()))

>              f.close()
>
>      if diff:
> -        template = parse_diff_for_commit_message(''.join(diff))
> +        template = parse_diff_for_commit_message(''.join(decode_list(diff)))
>
>      return template
>

marcus-h commented:
I think we are almost done:)

On 2019-07-19 00:06:15 -0700, Marco Strigl wrote:
> diff --git a/osc/commandline.py b/osc/commandline.py
> index 8ea37ba3..8b7bb952 100644
> --- a/osc/commandline.py
> +++ b/osc/commandline.py
> @@ -1405,16 +1405,19 @@ def _check_service(root):
>          rdiff = None
>          if opts.diff or not opts.message:
>              try:
> -                rdiff = 'old: %s/%s\nnew: %s/%s rev %s\n' % (dst_project, dst_package, src_project, src_package, rev)
> -                rdiff += decode_it(server_diff(apiurl,
> -                                    dst_project, dst_package, None,
> -                                    src_project, src_package, rev, True))
> +                rdiff = b'old: %s/%s\nnew: %s/%s rev %s\n' % (dst_project.encode(), dst_package.encode(), src_project.encode(), src_package.encode(), rev.encode())

Hmm rev could be None here. In the old code this was no "issue"
because it was simply interpolated as "None" but now it will
yield to an AttributeError.
To mimic the old behavior we could simply use str(rev).encode().

> diff --git a/osc/core.py b/osc/core.py
> index 23ca55f6..83cb7a1d 100644
> --- a/osc/core.py
> +++ b/osc/core.py
> @@ -4866,14 +4878,14 @@ def server_diff_noex(apiurl,
>          msg = None
>          body = None
>          try:
> -            body = decode_it(e.read())
> -            if not 'bad link' in body:
> -                return '# diff failed: ' + body
> +            body = e.read()
> +            if not b'bad link' in body:
> +                return b'# diff failed: ' + body
>          except:
> -            return '# diff failed with unknown error'
> +            return b'# diff failed with unknown error'
>
>          if expand:
> -            rdiff =  "## diff on expanded link not possible, showing unexpanded version\n"
> +            rdiff =  b"## diff on expanded link not possible, showing unexpanded version\n"

(We could remove the additional whitespace, but OK.)

>              try:
>                  rdiff += server_diff_noex(apiurl,
>                      old_project, old_package, old_revision,
> @@ -7241,19 +7253,20 @@ def get_commit_message_template(pac):
>      else:
>          todo = pac.filenamelist + pac.filenamelist_unvers
>
> +

Hmm?:)

>      files = [i for i in todo if i.endswith('.changes') and pac.status(i) in ('A', 'M')]
>
>      for filename in files:
>          if pac.status(filename) == 'M':
>              diff += get_source_file_diff(pac.absdir, filename, pac.rev)
>          elif pac.status(filename) == 'A':
> -            f = open(os.path.join(pac.absdir, filename), 'r')
> +            f = open(os.path.join(pac.absdir, filename), 'rb')
>              for line in f:
> -                diff += '+' + line
> +                diff.extend((b'+' + line for line in f))
>              f.close()
>
It will probably do the right thing but this looks a bit complicated.
What about:

with open(os.path.join(pac.absdir, filename), 'rb') as f:
    diff.extend((b'+' + line for line in f))

>      if diff:
> -        template = parse_diff_for_commit_message(''.join(diff))
> +        template = parse_diff_for_commit_message(''.join(decode_list(diff)))
>
>      return template
>

marcus-h commented:
Merged. Thanks a lot!:)
